docs(kic): fallback configuration guide

czeslavo opened the pull request:
### Description

Adds a guide explaining the new `FallbackConfiguration` feature usage.
 
Part of https://github.com/Kong/kubernetes-ingress-controller/issues/5934.


### Testing instructions

Preview link: https://deploy-preview-7463--kongdocs.netlify.app/kubernetes-ingress-controller/unreleased/guides/high-availability/fallback-config/

### Checklist 

- [x] Review label added <!-- (see below) -->
- [x] [Conditional version tags](https://docs.konghq.com/contributing/conditional-rendering/#conditionally-render-content-by-version) added, if applicable.

<!-- For example, if this change is for an upcoming 3.6 release, enclose your content in `{% if_version gte:3.6.x %} <content> {% endif_version %}` tags. 

Use any of the following keys:
* `gte:<version>` - greater than or equal to a specific version
* `lte:<version>` - less than or equal to a specific version
* `eq:<version>` - exactly equal to a specific version

You can do the same for older versions. -->

<!-- !!! Only Kong employees can add labels due to a GitHub limitation. If you're an OSS contributor, thank you! The maintainers will label this PR for you !!! -->

<!-- When raising a pull request, indicate what type of review you need with one of the following labels:

    review:copyedit: Request for writer review.
    review:general: Review for general accuracy and presentation. Does the doc work? Does it output correctly?
    review:tech: Request for technical review for a docs platform change.
    review:sme: Request for review from an SME (engineer, PM, etc).

At least one of these labels must be applied to a PR or the build will fail.
-->

pmalek commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
```diff
@@ -0,0 +1,327 @@
+
+In this example we'll demonstrate how the Fallback Configuration works in practice.
+
+{% include /md/kic/prerequisites.md release=page.release disable_gateway_api=false %}
```
This block seems to be expandable on click with a bit of a surprising behavior:

Freshly opened page:

<img width="1304" alt="image" src="https://github.com/Kong/docs.konghq.com/assets/739996/678c3edf-38ea-4a85-8a06-576d7858202a">

After clicking on the block:

<img width="1294" alt="image" src="https://github.com/Kong/docs.konghq.com/assets/739996/38101134-1296-4ea7-988a-b6e1890c3791">

\+ long explanation how to install Gateway API and Kong via helm chart. There is no clear indication where the expanded block ends and when the instructions from this page continue.

pmalek replied:
This also causes the TOC to be slightly misleading:

<img width="239" alt="image" src="https://github.com/Kong/docs.konghq.com/assets/739996/296b2dbb-d91f-42b7-b50f-15b4c043cb9f">

czeslavo replied:
It is a reusable building block to be used with `include` directive. You're right it doesn't look good, especially the TOC. I moved it to the beginning of the guide (similarly as it is done in other [guides](https://docs.konghq.com/kubernetes-ingress-controller/latest/guides/services/http/)) so it doesn't look weird.

mheap replied:
The styling is a pre-existing issue which is in the backlog

pmalek commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
```diff
@@ -0,0 +1,327 @@
+
+In the example, we'll consider a situation where:
+
+1. We have two `HTTPRoutes` pointing to the same `Service` and working correctly.
+2. One of the `HTTPRoutes` gets broken because of referencing an invalid `KongPlugin`.
```
Nit: (I'm not sure what's the convention in the docs around this)

```suggestion
1. We have two `HTTPRoute`s pointing to the same `Service` and working correctly.
2. One of the `HTTPRoutes` gets broken because of referencing an invalid `KongPlugin`.
```

pmalek commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
```diff
@@ -0,0 +1,327 @@
+
+{:.note}
+> Another way to monitor the Fallback Configuration mechanism is by Prometheus metrics. Please refer to the
+> [Prometheus Metrics](/production/observability/prometheus) for more information.
```
```suggestion
> [Prometheus Metrics](/kubernetes-ingress-controller/{{page.release}}/production/observability/prometheus) for more information.
```

pmalek commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
````diff
@@ -0,0 +1,327 @@
+We can verify this by inspecting the diagnostic endpoint:
+
+```bash
+kubectl port-forward -n kong deploy/kong-controller 8010:10256 &
````
Any particular reason to use a different port than already configured in the pod/container : `10256`?

czeslavo replied:
Not really, I'll change that to the pod's one.

pmalek reviewed (commented):
First round of review (without taking this for a spin):

- I'd add section to this doc so that it's easier to follow. Currently this only has `Example scenario` which doesn't tell the reader what stage of the scenario they're at (am I deploying manifests, testing them, patching them or verifying that fallback config works?)
- I'd dd a separate (named) section which shows how last known config backfill + fallback works on concrete example. This should be separate in my eyes from what already described and the difference should be noted (`X` route still works, having previous config filled in).

pmalek commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
````diff
@@ -0,0 +1,327 @@
+
+Create a `KongPlugin` with an invalid configuration:
+
+```bash
+echo 'apiVersion: configuration.konghq.com/v1
+kind: KongPlugin
+metadata:
+  name: key-auth
+plugin: key-auth
+config:
+  # Should be 'key_names', not 'keys'.
+  keys: ["key"]' | kubectl apply -f -
+```
````
This plugin fails admission webhook validation:

```
Error from server: error when creating kubernetes-ingress-controller/examples/gateway-httproute.yaml": admission webhook "validations.kong.konghq.com" denied the request: plugin failed schema validation: schema violation (config.keys: unknown field)
```

Don't we want to provide an example that will pass that stage and fail during Gateway configuration apply phase (as noted in the `How it works` section above, point 3)?

czeslavo replied:
Yeah, that's why I added the step to disable the webhook with an explanation saying it will allow us to slip invalid configuration to stage 3. As we discussed once, I think it will be more _reliable_ to demonstrate it this way (turning the webhook off and pushing something that the webhook would reject). If we knew there was a problem that we could trigger with the webhook turned on, we should rather plumb it in the webhook or the translation phase, not make it an example in the guide, right? 😄

pmalek replied:
👍 fair point.

mheap replied:
I came here to say this. We should provide an example that can actually happen (such as duplicate plugins on a single service) rather than disabling the admission webhook

czeslavo replied:
@mheap My rationale for doing it this way was that by disabling the webhook server we can reliably showcase this feature. If we have any known issue, we tend to fix that either in the webhook or the translation as soon as possible, therefore using a known issue as an example doesn't make much sense. In fact, duplicate plugins on a single service have already been covered in the webhook:
```
Error from server: admission webhook "validations.kong.konghq.com" denied the request: Service has invalid KongPlugin annotation: cannot attach multiple plugins: key-auth-1, key-auth-2 of the same type key-auth
```

programmer04 replied:
Just for broader context, it doesn't work every time see
- https://github.com/Kong/kubernetes-ingress-controller/issues/5995

and this is connected with the behavior described in
- https://github.com/Kong/kubernetes-ingress-controller/issues/6139 

so I think that in the future we'll cover it with webhook too

mheap replied:
I get an event, but it isn't blocked by the admission controller for me:

```
default       0s          Warning   KongConfigurationApplyFailed     kongplugin/rate-limiting-authenticated         invalid plugin:rate-limiting: uniqueness violation: 'plugins' entity with primary key set to '35619164-260d-5ac2-a840-b86e6cc87f76' already declared
```

```
❯ kaf .
service/echo created
deployment.apps/echo created
ingress.networking.k8s.io/echo created
secret/alex-key-auth created
kongconsumer.configuration.konghq.com/alex created
secret/bob-key-auth created
kongconsumer.configuration.konghq.com/bob created
kongplugin.configuration.konghq.com/key-auth-example created
kongplugin.configuration.konghq.com/rate-limiting-base created
kongplugin.configuration.konghq.com/rate-limiting-authenticated created
```

Notice the commented out annotation on the `alex` consumer, which makes the rate limit plugin on the `echo` service a duplicate:

```
❯ kaf .
service/echo created
deployment.apps/echo created
ingress.networking.k8s.io/echo created
secret/alex-key-auth created
kongconsumer.configuration.konghq.com/alex created
secret/bob-key-auth created
kongconsumer.configuration.konghq.com/bob created
kongplugin.configuration.konghq.com/key-auth-example created
kongplugin.configuration.konghq.com/rate-limiting-base created
kongplugin.configuration.konghq.com/rate-limiting-authenticated created
```

```
apiVersion: v1
kind: Service
metadata:
  labels:
    app: echo
  name: echo
  annotations:
    konghq.com/plugins: key-auth-example, rate-limiting-base, rate-limiting-authenticated
spec:
  ports:
    - port: 1025
      name: tcp
      protocol: TCP
      targetPort: 1025
    - port: 1026
      name: udp
      protocol: TCP
      targetPort: 1026
    - port: 1027
      name: http
      protocol: TCP
      targetPort: 1027
  selector:
    app: echo
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: echo
  name: echo
spec:
  replicas: 1
  selector:
    matchLabels:
      app: echo
  strategy: {}
  template:
    metadata:
      labels:
        app: echo
    spec:
      containers:
        - image: kong/go-echo:latest
          name: echo
          ports:
            - containerPort: 1027
          env:
            - name: NODE_OWNER
              value: Michael
            - name: NODE_NAME
              valueFrom:
                fieldRef:
                  fieldPath: spec.nodeName
            - name: POD_NAME
              valueFrom:
                fieldRef:
                  fieldPath: metadata.name
            - name: POD_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
            - name: POD_IP
              valueFrom:
                fieldRef:
                  fieldPath: status.podIP
          resources: {}
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: echo
  annotations:
    konghq.com/strip-path: "true"
spec:
  ingressClassName: kong
  rules:
    - http:
        paths:
          - path: /echo
            pathType: ImplementationSpecific
            backend:
              service:
                name: echo
                port:
                  number: 1027
---
apiVersion: v1
kind: Secret
metadata:
  name: alex-key-auth
  labels:
    konghq.com/credential: key-auth
stringData:
  key: hello_world
---
apiVersion: configuration.konghq.com/v1
kind: KongConsumer
metadata:
  name: alex
  annotations:
    kubernetes.io/ingress.class: kong
    # konghq.com/plugins: rate-limiting-authenticated
username: alex
credentials:
  - alex-key-auth
---
apiVersion: v1
kind: Secret
metadata:
  name: bob-key-auth
  labels:
    konghq.com/credential: key-auth
stringData:
  key: bob_password
---
apiVersion: configuration.konghq.com/v1
kind: KongConsumer
metadata:
  name: bob
  annotations:
    kubernetes.io/ingress.class: kong
username: bob
credentials:
  - bob-key-auth
---
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: key-auth-example
plugin: key-auth
config:
  key_names:
    - apikey
---
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: rate-limiting-base
plugin: rate-limiting
config:
  second: 1
  policy: local
---
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: rate-limiting-authenticated
plugin: rate-limiting
config:
  second: 5
  policy: local
```

czeslavo replied:
@mheap I've rewritten the guide building on top of your example, slightly modifying it by using two `HTTPRoute`s and duplicating plugins on one of the routes. Please take a look: https://github.com/Kong/docs.konghq.com/pull/7463/commits/4ef76e251e6e8e6761fbab8d091792a57d90a1b8

czeslavo commented:
@pmalek I noticed that the `/debug/config/problems` endpoint doesn't work as expected - giving you a heads up so you don't waste time figuring out why it doesn't give you proper responses.

czeslavo commented:
> First round of review (without taking this for a spin):
> 
> * I'd add section to this doc so that it's easier to follow. Currently this only has `Example scenario` which doesn't tell the reader what stage of the scenario they're at (am I deploying manifests, testing them, patching them or verifying that fallback config works?)
> * I'd dd a separate (named) section which shows how last known config backfill + fallback works on concrete example. This should be separate in my eyes from what already described and the difference should be noted (`X` route still works, having previous config filled in).

Added more granular sections as well as the whole section about backfilling. Thanks for the comments, I hope I addressed all of them. 🙇

pmalek commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
````diff
@@ -0,0 +1,476 @@
+Let's detach the broken `KongPlugin` from `route-b` so we get an entirely valid configuration:
+
+```bash
+kubectl annotate --overwrite httproute route-b konghq.com/plugins=
````
```suggestion
kubectl annotate --overwrite httproute route-b konghq.com/plugins-
```

I believe we want to remove the annotation rather than set it to empty string.

pmalek commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
````diff
@@ -0,0 +1,476 @@
+Using diagnostic endpoints, we can now inspect the objects that were excluded and backfilled in the configuration:
+
+```bash
+kubectl port-forward -n kong deploy/kong-controller 10256:10256 &
````
Nit: 

```suggestion
kubectl port-forward -n kong deploy/kong-controller 10256 &
```

pmalek commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
````diff
@@ -0,0 +1,476 @@
+We can verify this by inspecting the diagnostic endpoint:
+
+```bash
+kubectl port-forward -n kong deploy/kong-controller 10256:10256 &
````
```suggestion
kubectl port-forward -n kong deploy/kong-controller 10256 &
```

pmalek commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
````diff
@@ -0,0 +1,476 @@
+kubectl annotate httproute route-b konghq.com/plugins=key-auth
+```
+
+The results should look like this:
+```text
````
Nit: inconsistent spacing between text and code blocks

```suggestion
The results should look like this:

```text
```

pmalek commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
````diff
@@ -0,0 +1,476 @@
+  keys: ["key"]' | kubectl apply -f -
+```
+
+The results should look like this:
+```text
````
```suggestion
The results should look like this:

```text
```

pmalek commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
````diff
@@ -0,0 +1,476 @@
+curl -i $PROXY_IP/route-b
+```
+
+The results should look like this:
+```text
````
```suggestion
The results should look like this:

```text
```

pmalek commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
````diff
@@ -0,0 +1,476 @@
+curl -i localhost:10256/debug/config/problems
+```
+
+The results should look like this:
+```text
````
```suggestion
The results should look like this:

```text
```

pmalek replied:
And ditto in other places.

pmalek reviewed (commented):
Looks good 👍 

Can we already add docs for the feature gate at https://github.com/Kong/docs.konghq.com/blob/86f6865fae98afe729d77583563d982060465578/app/_src/kubernetes-ingress-controller/reference/feature-gates.md ?

czeslavo commented:
> Can we already add docs for the feature gate at https://github.com/Kong/docs.konghq.com/blob/86f6865fae98afe729d77583563d982060465578/app/_src/kubernetes-ingress-controller/reference/feature-gates.md ?

Updated the feature gates reference (we were missing some from KIC 3.1): https://github.com/Kong/docs.konghq.com/pull/7463/commits/52882287539686a0facfe2522ace96a860d92f3b

mheap commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
```diff
@@ -0,0 +1,480 @@
+
+## Overview 
+
+{{site.kic_product_name}} in version 3.2.0 introduced the Fallback Configuration
```
```suggestion
{{site.kic_product_name}} 3.2.0 introduced the Fallback Configuration
```

czeslavo replied:
https://github.com/Kong/docs.konghq.com/pull/7463/commits/36b0389f32ab2fcc73257894afe6e40cb80c77c1

mheap commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
```diff
@@ -0,0 +1,480 @@
+
+{:.note}
+> **Note:** The Fallback Configuration is an opt-in feature. You must
+> enable it by setting `FallbackConfiguration=true` in the controller's feature
```
Can `enable it` be a link to the docs on how to do so?

czeslavo replied:
https://github.com/Kong/docs.konghq.com/pull/7463/commits/36b0389f32ab2fcc73257894afe6e40cb80c77c1

mheap commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
```diff
@@ -0,0 +1,480 @@
+- Automatically builds a fallback configuration that Kong will accept without requiring user intervention by
+  either:
+  - Excluding the broken objects along with its dependants.
+  - Backfilling the broken object along with its dependants using the last valid Kubernetes objects' in-memory cache (if `--use-last-valid-config-for-fallback` flag is set).
```
Most people use env variables to configure in k8s. Should we show `CONTROLLER_USE_LAST_VALID_CONFIG_FOR_FALLBACK` rather than the flag?

czeslavo replied:
https://github.com/Kong/docs.konghq.com/pull/7463/commits/36b0389f32ab2fcc73257894afe6e40cb80c77c1

mheap commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
```diff
@@ -0,0 +1,480 @@
+
+| `FallbackConfiguration` feature gate value | `--use-last-valid-config-for-fallback` flag value | Behavior                                                                                                                                             |
+|--------------------------------------------|---------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `false`                                    | `false`/`true` (has no effect)                    | The last valid configuration is used as a whole to recover (if stored).                                                                              |
```
Should KIC throw an error if `--use-last-valid-config-for-fallback` is passed when `FallbackConfiguration` is false?

czeslavo replied:
Yeah, that makes sense. Posted a PR adding that: https://github.com/Kong/kubernetes-ingress-controller/pull/6168

mheap commented on app/_src/kubernetes-ingress-controller/guides/high-availability/fallback-config.md:
```diff
@@ -0,0 +1,480 @@
+
+Below diagram illustrates how the Fallback Configuration feature works in detail:
+
+![fallback-config](/assets/images/products/kubernetes-ingress-controller/fallback-config.png "Fallback Configuration diagram")
```
Should this be a mermaid diagram instead? cc @lena-larionova

czeslavo replied:
I gave Mermaid a go although I wouldn't say I like the results much 😅 Probably this could be styled to look more appealing:

```mermaid
flowchart TD
classDef sub opacity:0
classDef note fill:#ffd, stroke:#ccb
classDef externalCall fill:#9e8ebf
classDef startEnd fill:#545454

    A("`Update loop triggered`") --> B["`Generate Kubernetes objects' store snapshot to be passed to the Translator`"]
    B --> C["`Translator: generate Kong configuration based on the generated snapshot`"]
    C --> D["`Configure Kong Gateways using generated declarative configuration`"]
    D --> E{"`Configuration rejected?`"}
    E --> |No| G["`Store the Kubernetes objects' snapshot to be used as the last valid state`"]
    E --> |Yes| F["`Build a dependency graph of Kubernetes objects - using the snapshot`"]
    G --> H["`Store the declarative configuration to be used as the last valid configuration`"]
    H --> Z("`End of the loop`")
    F --> I["`Exclude an object along with all its dependants from the fallback Kubernetes objects snapshot`"]
    I --> J["`Add a previous valid version of the object along with its dependants' previous versions to the fallback snapshot`"]
    J --> K["`Translator: generate Kong configuration based on the fallback snapshot`"]
    K --> L["`Configure Kong Gateways using generated fallback declarative configuration`"]
    L --> M{"`Fallback configuration rejected?`"}
    M --> |Yes| N{"`Was the last valid configuration preserved?`"}
    N --> |Yes| O["`Configure Kong Gateways using the last valid declarative configuration`"]
    O --> Z
    N --> |No| Z
    M --> |No| P["`Store the fallback Kubernetes objects' snapshot to be used as the last valid state`"]
    P --> R["`Store the fallback declarative configuration to be used as the last valid configuration`"]

    subgraph subI [" "]
        I
        noteI[For every invalid object reported by the Gateway]
    end

    subgraph subJ [" "]
        J
        noteJ[Given there was a last valid Kubernetes objects' store snapshot preserved and the object is present]
    end

    class subI,subJ sub
    class noteI,noteJ note
    class D externalCall
    class A,Z startEnd
```

mheap replied:
Have you tried it in the docs site? We have custom styling

lena-larionova replied:
It should, though it might be complicated to make all the shapes go in the right order. I can help turn this into mermaid.js format

czeslavo replied:
> Have you tried it in the docs site? We have custom styling
Yup, I've tried. Didn't look well to my eyes. 😢

czeslavo replied:
@lena-larionova I believe you can start from what I posted in the comment above, there's a copy button in the top-right corner of the rendered diagram).

lena-larionova replied:
* upside: I got the thing styled and managed to have some of the nodes take up less space. Added it to the PR to test
* downside: The diagram is still massive, and I think we're close to the limit of how much smaller we can force it.

Honestly, I prefer the png condensed version of the diagram, but ideally we use mermaid for maintainability's sake.

czeslavo replied:
@lena-larionova Cannot argue - the mermaid variant still takes more space on display than png, but as you mentioned, it's better for maintainability's sake and I think it's more important. BTW the styling looks really cool, thanks for fixing that! 😄

mheap reviewed (approved):
This is an awesome guide, thank you.

The config debug endpoint isn't working (known issue), but otherwise it all worked as expected (with my latest commit to fix Helm values)

czeslavo commented:
Labeling `do not merge` as we need to:
- decide whether to keep the Mermaid diagram or the PNG one
- update the responses to align with what will be shipped after https://github.com/Kong/kubernetes-ingress-controller/pull/6159 gets merged
